lexos/models/kmeans_model.py

```diff
@@ -0,0 +1,475 @@
+It uses sklearn.cluster.KMeans for most important analysis, please see:
+http://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
+"""
+
+import numpy as np
+import pandas as pd
+import colorlover as cl
+import plotly.graph_objs as go
+from plotly.offline import plot
+from sklearn.decomposition import PCA
+from sklearn.cluster import KMeans as KMeans
+from typing import Optional, NamedTuple, List
+from lexos.models.base_model import BaseModel
+from lexos.models.matrix_model import MatrixModel
+from lexos.receivers.matrix_receiver import IdTempLabelMap
+from lexos.helpers.error_messages import EMPTY_DTM_MESSAGE
+from lexos.receivers.kmeans_receiver import KMeansOption, KMeansReceiver, \
+    KMeansViz
+
+# Alias for typed tuple to increase readability.
+PlotlyHTMLPlot = str
+HTMLTable = str
+
+
+class KMeansResult(NamedTuple):
+    """A typed tuple to hold processed k-means results."""
+
+    plot: PlotlyHTMLPlot
+    table: HTMLTable
+
+
+class KMeansTestOptions(NamedTuple):
+    """A typed tuple to hold k-means test options."""
+
+    doc_term_matrix: pd.DataFrame
+    id_temp_label_map: IdTempLabelMap
+    front_end_option: KMeansOption
+
+
+class KMeansUnprocessedResult(NamedTuple):
+    """A typed tuple to hold unprocessed k-means results."""
+
+    plot: go.Figure
+    table: pd.DataFrame
+
+
+class KMeansModel(BaseModel):
+    """This is the class to run k-means analysis.
+
+    :param test_options: the input used in testing to override the
+                         dynamically loaded option.
+    """
+
+    def __init__(self, test_options: Optional[KMeansTestOptions] = None):
+        """Initialize the class based if test option is given."""
+        super().__init__()
+        if test_options is not None:
+            self._test_dtm = test_options.doc_term_matrix
+            self._test_front_end_option = test_options.front_end_option
+            self._test_id_temp_label_map = test_options.id_temp_label_map
+        else:
+            self._test_dtm = None
+            self._test_front_end_option = None
+            self._test_id_temp_label_map = None
+
+    @property
+    def _doc_term_matrix(self) -> pd.DataFrame:
+        """:return: the document term matrix."""
+        return self._test_dtm if self._test_dtm is not None \
+            else MatrixModel().get_matrix()
+
+    @property
+    def _id_temp_label_map(self) -> IdTempLabelMap:
+        """:return: a map takes an id to temp labels."""
+        return self._test_id_temp_label_map \
+            if self._test_id_temp_label_map is not None \
+            else MatrixModel().get_id_temp_label_map()
+
+    @property
+    def _k_means_front_end_option(self) -> KMeansOption:
+        """:return: a typed tuple that holds the k-means front end option."""
+        return self._test_front_end_option \
+            if self._test_front_end_option is not None \
+            else KMeansReceiver().options_from_front_end()
+
+    def _get_reduced_data(self) -> np.ndarray:
```

Comment:
If you have time I think you need to think about refactoring this method, since this method either returns a 2D array or a 3D array, which is two different data type.

Reply:
If you don't have time, leave it like this. This is probably fine.